sdk/python/kfp/components/structures.py

```diff
@@ -231,6 +233,25 @@ def transform_env(self) -> None:
         """Use None instead of empty dict for env."""
         self.env = None if self.env == {} else self.env
 
+    def validate_command(self) -> None:
+        """Validate command doesn't contain artifact channel."""
+        self._validate_no_artifact_channel(self.command)
+
+    def validate_args(self) -> None:
+        self._validate_no_artifact_channel(self.args)
+
+    def validate_env(self) -> None:
```

Comment:
I wasn't able to get a pipeline that uses an env var placeholder to compile. I think it's possible that env vars cannot be placeholders and that the type annotation for `ContainerSpec.env` is wrong and should be `Optional[Mapping[str, str]]` (or better yet `Optional[Dict[str, str]]` because we really only test on dictionaries, not arbitrary mappings).

WDYT? Are you able to get an env var with a placeholder value to compile?

Reply:
Yes, thanks for the note! I am also not able to get a env var with placeholder to compile and will remove the `validate_env` method for now.